### PR TITLE
Add CatalogVersion to fix the incompatibility issue

### PR DESCRIPTION
**Pull request by weiqingy**
## What changes were proposed in this pull request?

1. Add `CatalogVersion` to fix the incompatibility issue
2. Add an example

## How was this patch tested?
Pass the current unit tests.

### CODE REVIEW

**Comment by weiqingy** [comment]
Hi, @dongjoon-hyun @bikassaha Could you please help to review this PR? Thanks. cc: @merlintang

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 1]
```diff
@@ -30,6 +30,31 @@ import org.apache.spark.sql.catalyst.util.DataTypeParser
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.execution.datasources.hbase.types._
 
+case class CatalogVersion(major: Int, minor: Int) extends Comparable[CatalogVersion] {
+  override def compareTo(o: CatalogVersion): Int = {
+    if(major > o.major)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: nit. `if(` -> `if (`?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 2]
```diff
@@ -30,6 +30,31 @@ import org.apache.spark.sql.catalyst.util.DataTypeParser
...
+      1
+    else if(major == o.major)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: here, too.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 3]
```diff
@@ -30,6 +30,31 @@ import org.apache.spark.sql.catalyst.util.DataTypeParser
...
+    else
+      -1
+  }
+
+  override def toString: String = major + "." + minor
+}
+
+object CatalogVersion {
+  def apply(s: String): CatalogVersion = {
+    if(!s.matches("^[0-9]*\\.[0-9]*$"))
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: `if(` -> `if (`.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 4]
```diff
@@ -30,6 +30,31 @@ import org.apache.spark.sql.catalyst.util.DataTypeParser
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.execution.datasources.hbase.types._
 
+case class CatalogVersion(major: Int, minor: Int) extends Comparable[CatalogVersion] {
+  override def compareTo(o: CatalogVersion): Int = {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: What if `o` is null? Can we implement this as null-safe? Otherwise, NPE might occur.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 3 above)
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: This seems to match `dot` string; `.`. Then, parse error might occurs on line 52~53.
Could we change `^[0-9]*\\.[0-9]*$` into `^[0-9]+\\.[0-9]+$`?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 5]
```diff
@@ -206,6 +231,8 @@ object HBaseTableCatalog {
   val length = "length"
   val fCoder = "coder"
   val tableCoder = "tableCoder"
+  // The version num of catalog
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: nit. `num` -> `number`.

File: core/src/test/scala/org/apache/spark/sql/PhoenixSuite.scala [hunk 6]
```diff
@@ -61,7 +61,7 @@ object PRecord {
 
 class PhoenixSuite extends SHC with Logging {
   override def catalog = s"""{
-                    |"table":{"namespace":"default", "name":"table1", "tableCoder":"Phoenix"},
+                    |"table":{"namespace":"default", "name":"table1", "tableCoder":"Phoenix", "version":"2.0"},
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/sql/PhoenixSuite.scala]: I'm just wondering. Is this okay? It looks too great change.
Next time, we need `3.0`?

File: examples/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/DataCoder.scala [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+ */
+
+package org.apache.spark.sql.execution.datasources.hbase.examples
+
+import org.apache.spark.sql.execution.datasources.hbase._
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.sql.{SQLContext, DataFrame}
+
+case class DCRecord(col00: String,
+                   col01: Int,
```
Comment (dongjoon-hyun) [inline examples/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/DataCoder.scala]: nit, alignment?

File: examples/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/DataCoder.scala [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                |"col4":{"cf":"CF2", "col":"COL4", "type":"int"},
+                |"col5":{"cf":"CF3", "col":"COL5", "type":"bigint"},
+                |"col6":{"cf":"CF3", "col":"COL6", "type":"smallint"},
+                |"col7":{"cf":"CF4", "col":"COL7", "type":"string"},
+                |"col8":{"cf":"CF4", "col":"COL8", "type":"tinyint"}
+                |}
+                |}""".stripMargin
+
+  def main(args: Array[String]){
+    val sparkConf = new SparkConf().setAppName("CompositeKeyTest")
```
Comment (dongjoon-hyun) [inline examples/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/DataCoder.scala]: Since this is inside `example`, can we use `CompositeKeyExample` here?

File: examples/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/DataCoder.scala [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+    def withCatalog(cat: String): DataFrame = {
+      sqlContext
+        .read
+        .options(Map(HBaseTableCatalog.tableCatalog->cat))
+        .format("org.apache.spark.sql.execution.datasources.hbase")
+        .load()
+    }
+
+    //populate table with composite key
```
Comment (dongjoon-hyun) [inline examples/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/DataCoder.scala]: nit. `// populate`

File: examples/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/DataCoder.scala [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    val data = (0 to 255).map { i =>
+      DCRecord(i)
+    }
+
+    sc.parallelize(data).toDF.write.options(
+      Map(HBaseTableCatalog.tableCatalog -> cat, HBaseTableCatalog.newTable -> "5"))
```
Comment (dongjoon-hyun) [inline examples/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/DataCoder.scala]: indentation?

**Comment by weiqingy** [comment]
Thanks for the quick feedback, @dongjoon-hyun . I will update the PR soon.

File: README.md [hunk 11]
```diff
@@ -1,20 +1,17 @@
...
 
 Note that if user want dataframe to only handle byte array, the binary type can be specified. Then user can get the catalyst row with each column as a byte array. User can further deserialize it with customized deserializer, or operate on the RDD of the data frame directly.
 
-## Solve the conflicts of Catalyst datatype order and HBase byte array order
-The libary automatically handle the orderness between the conflicts between the java data type (Short, Integer, Long, Float, Double, String, etc) and HBase bytearray order, which means as long as the format saved in HBase is consistent with the Java native data format, the libary will be able to take care of pruning and comparison instead of relying on a specific import tools.
```
Comment (bikassaha) [inline README.md]: Why is this removed?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 12]
```diff
@@ -30,6 +30,31 @@ import org.apache.spark.sql.catalyst.util.DataTypeParser
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.execution.datasources.hbase.types._
 
+case class CatalogVersion(major: Int, minor: Int) extends Comparable[CatalogVersion] {
```
Comment (bikassaha) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: Major and minor version is probably an overkill.

File: core/src/test/scala/org/apache/spark/sql/PhoenixSuite.scala (hunk 6 above)
Comment (bikassaha) [inline core/src/test/scala/org/apache/spark/sql/PhoenixSuite.scala]: Since we just added this, it looks weird to start with 2.0 instead of 1.0

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 13]
```diff
@@ -220,8 +247,20 @@ object HBaseTableCatalog {
...
-    // If 'tableCoder' is not specified in the catalog, 'Phoenix' will be the table coder
-    val tCoder = tableMeta.getOrElse(tableCoder, SparkHBaseConf.Phoenix).asInstanceOf[String]
+
+    // The snippet below is to fix the incompatibility issue caused by the change of adding
+    // Phoenix as coder. Before the change, the catalog version is 1.0. If the catalog
+    // version specified by user is later than 1.0, tableCoder must be specified too.
+    // The default value of the catalog version is 1.0. When the catalog version is 1.0,
+    // the default value of tableCoder is PrimitiveType.
+    val vNum = tableMeta.getOrElse(cVersion, "1.0").asInstanceOf[String]
+    if (CatalogVersion(vNum).compareTo(CatalogVersion("1.0")) == 1){
```
Comment (bikassaha) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: We may be using version often if we keep making changes to the catalog def. So would be good to have an int representation that can be easily compared without having to do string manipulation. If we drop minor version then it may be straightforward to do.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 14]
```diff
@@ -220,8 +247,20 @@ object HBaseTableCatalog {
...
+      if (tableMeta.get(tableCoder).isEmpty)
+        throw new UnsupportedClassVersionError("Please specify 'tableCoder' in your catalog " +
```
Comment (bikassaha) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: UnsupportedClassVersionError? These have specific meanings in Java. So its good to not use them outside of those well defined meanings. In this case we could use InvalidParameterException or something similar.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 15]
```diff
@@ -220,8 +247,20 @@ object HBaseTableCatalog {
...
+          "if the catalog version is later than 1.0")
+    }
+
+    val tCoder = tableMeta.getOrElse(tableCoder, SparkHBaseConf.PrimitiveType).asInstanceOf[String]
```
Comment (bikassaha) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: IMO the safe way to write this kind of logic would be
if (version == 1.0) { tableCoder = "primitive" } else { tableCoder = tableMeta.getOrElseThrowException } 

The way its currently written tCoder = tableMeta.getOrElse() is outside the version logic but dependent on it. And in future, someone could change the version logic and leave this line unchanged leading to bugs.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 13 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: I understand using a single major version would be much simpler. It's just a little weird to me. Nowadays people have got used to version numbers and all versions are at least like "x.y", from iOS, Windows, and open source software. What's more, if "1" is a valid version number, user would probably expect "1.0" valid.

Still, I'd prefer to handle version in a more general way, leaving room for future possible minor version upgrades. Java 9 is going to have its own Version class and we probably just replace this version class with the Java one then.

File: core/src/test/scala/org/apache/spark/sql/PhoenixSuite.scala (hunk 6 above)
Comment (weiqingy) [inline core/src/test/scala/org/apache/spark/sql/PhoenixSuite.scala]: This version is the one with "Phoenix" feature in. I suppose it should be 2.0 so the previous version (without Phoenix support) could be 1.0?

File: README.md (hunk 11 above)
Comment (weiqingy) [inline README.md]: This is just for PrimitiveType. PrimitiveType invokes HBase APi `Bytes.toBytes` to encode data, so it needs to do extra work to handle the conflicts of byte array order between java data type and HBase bytearray order. Now SHC has added Phoenix as coder, which relies on Phoenix data types to encode/decode data.

File: core/src/test/scala/org/apache/spark/sql/PhoenixSuite.scala (hunk 6 above)
Comment (weiqingy) [inline core/src/test/scala/org/apache/spark/sql/PhoenixSuite.scala]: This version 2.0 is the one with "Phoenix" feature in. I suppose the previous version (without Phoenix support) could be 1.0.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 4 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: I am actually neutral about this - I noticed in the Java library source code, they don't check whether the parameter of compareTo is null or not, like in Integer.compareTo, String.compareTo. They just choose to let NPE occur. I suppose maybe compareTo is a so widely used method and everybody knows a null object should not be passed, and if it does happen, a NPE suffices.

I am personally OK with either: NPE or IllegalArgumentException.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 1 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: Done.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 2 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: Done.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 3 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: I have updated the PR for this.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 14 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: I created a class `CatalogDefinitionException` for this.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 15 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: Thanks. I have updated the PR for this.

**Comment by dongjoon-hyun** [comment]
Is it ready for review?

**Comment by weiqingy** [comment]
@dongjoon-hyun Yes, it's ready.

**Comment by dongjoon-hyun** [comment]
Could you resolve the conflicts?

**Comment by weiqingy** [comment]
@dongjoon-hyun The conflicts have been fixed.

**Comment by dongjoon-hyun** [comment]
LGTM

**Comment by weiqingy** [comment]
Thanks for the review, @dongjoon-hyun .
